soroban-env-host/src/host.rs

```diff
@@ -799,9 +818,59 @@ impl Host {
             // maintains a borrow of self.0.contracts, which can cause borrow errors.
             let cfs_option = self.0.contracts.borrow().get(&id).cloned();
             if let Some(cfs) = cfs_option {
-                return self.with_frame(Frame::TestContract(id.clone(), func.clone()), || {
-                    cfs.call(&func, self, args)
-                        .ok_or_else(|| self.err_general("function not found"))
+                let frame = TestContractFrame::new(id.clone(), func.clone());
```

Comment:
❔ Does this change the behavior at all of a `Env::call` in a panic scenario? I don't think so, but not sure as I see changes in this `call_n` function.

Reply:
I'm not 100% sure what you're asking. The .. whole point is to change behaviour! I'll try to answer by detailing the before and after behaviour and you can tell me if it does or doesn't meet expectations.

First: this only affects native contracts and calls between them. No change to VM-based calls.

Before this change:
  - If contract A `call`s contract B and B panics, the panic will unwind through the host (actually somewhat corrupting the host by failing to `rollback` the `TestContractFrame`) and all the way back to A. If A happens to have `catch_panic` or something they might catch it, otherwise it bubbles up to the outermost process catch handler which will probably print the panic.
  - If contract A `try_call`s contract B and B panics, the panic also bubbles up, `try_call` doesn't contain it any better.
  - If the panic originated from `impl Env for CheckedEnv` unwrapping a `HostError` from a failed host function called by B, the panic will contain a stringification of that `HostError` which captures where the error originates, as well as (I think?) a duplicate stack trace of where the `unwrap` failed (moments after the `HostError` was captured). 

After this change
  - If contract A `call`s contract B and B panics, the panic will unwind to `Host::call_n` and be caught and converted to a `HostError`. The host will `rollback` the `TestContractFrame` and then return `Err(HostError)` to `call`, which will return that `Err(HostError)` to `impl Env for CheckedEnv`, which will call back to `Host::escalate_error_to_panic`, which will panic and unwind into A, where its fate will be the same as the panic in the before-case.
  - If contract A `try_call`s contract B and B panics, the same initial sequence of events happens but when `Err(HostError)` is returned from `call_n` to `try_call`, it will be converted to `Ok(status)` which will pass through `impl Env for CheckedEnv` and return to B as a `RawVal`-that-happens-to-be-a-`Status`. No further unwinding happens, `try_call` contained the panic.
  - If the panic originated from `impl Env from CheckedEnv` escalating a `HostError` from a failed host function called by B, the panic's originating context (at least its stack trace and captured error buffer) will _mostly_ be eaten by `call_n` catching it. Its status will be propagated to the `HostError` returned from `call_n`

Reply:
Got it. My question was directed at wanting to make sure the Env::call behavior wasn't changing significantly, or if it was, understanding the change. I think this sounds right as you describe it.

In some ways this erode the Rust testing experience because someone chucking a panic may expect their debugger to take them to exactly where the original panic occurred, but they won't have that experience anymore. I think we have to balance creating that ideal experience vs the ideal same experience as production, and I think this PR makes the right tradeoff.

Reply:
I checked this! As it happens, the vscode debugger (by default) sets a breakpoint on the panic-initiating site, so at the moment of the first panic you break into a debugger.

Reply:
Oh nice. Yay!